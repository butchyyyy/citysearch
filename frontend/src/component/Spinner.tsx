import React from "react"

import "style/Spinner.less"

interface Props {
  /** If true, spinner animation and blocking mask will show over the wrapped elements */
  loading: boolean
  /** Wrapped elements to show spinner and blocking mask on when loading === true */
  children?: any
}

/**
 * UX class that shows spinner animation when e.g. an async action is fetching data
 */
const Spinner = (props: Props) => (
    <>
      {props.loading &&
      <div className="spinner">
        <div className="rect1" />
        <div className="rect2" />
        <div className="rect3" />
        <div className="rect4" />
        <div className="rect5" />
      </div>}
      <div className={props.loading ? "mask" : null}>{props.children}</div>
    </>
)

export default Spinner
