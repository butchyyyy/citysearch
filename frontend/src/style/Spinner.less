@spinner-width: 50px;
@spinner-height: 30px;

.mask {
  z-index: 100000;
  opacity: 0.3;
  background-color: white;
  pointer-events: none;
}

.spinner {
  width: @spinner-width;
  height: @spinner-height;
  text-align: center;
  font-size: 10px;
  left: 50%;
  top: 50%;
  margin-left: -@spinner-width/2;
  margin-top: -@spinner-height/2;
  position: absolute;
  z-index: 100001;
}

.spinner > div {
  background-color: #333;
  margin: 1px;
  height: 100%;
  width: 6px;
  display: inline-block;

  -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
  animation: sk-stretchdelay 1.2s infinite ease-in-out;
}

.spinner .rect1 {
  -webkit-animation-delay: -1.2s;
  animation-delay: -1.2s;
}

.spinner .rect2 {
  -webkit-animation-delay: -1.1s;
  animation-delay: -1.1s;
}

.spinner .rect3 {
  -webkit-animation-delay: -1.0s;
  animation-delay: -1.0s;
}

.spinner .rect4 {
  -webkit-animation-delay: -0.9s;
  animation-delay: -0.9s;
}

.spinner .rect5 {
  -webkit-animation-delay: -0.8s;
  animation-delay: -0.8s;
}

@-webkit-keyframes sk-stretchdelay {
  0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
  20% { -webkit-transform: scaleY(1.0) }
}

@keyframes sk-stretchdelay {
  0%, 40%, 100% {
    transform: scaleY(0.4);
    -webkit-transform: scaleY(0.4);
  }  20% {
       transform: scaleY(1.0);
       -webkit-transform: scaleY(1.0);
     }
}