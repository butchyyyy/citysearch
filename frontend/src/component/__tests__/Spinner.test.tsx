import Spinner from "component/Spinner"
import { shallow } from "enzyme"
import React from "react"

describe("Test Spinner Component", () => {
  const wrapper = shallow(<Spinner loading={false}>Children</Spinner>)

  it("Test Spinner Snapshot", () => {
    expect(wrapper).toMatchSnapshot()
    wrapper.setProps({ loading: true })
    expect(wrapper).toMatchSnapshot()
  })
})
