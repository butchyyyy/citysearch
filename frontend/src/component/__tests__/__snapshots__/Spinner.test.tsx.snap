// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Spinner Component Test Spinner Snapshot 1`] = `
<Fragment>
  <div
    className={null}
  >
    Children
  </div>
</Fragment>
`;

exports[`Test Spinner Component Test Spinner Snapshot 2`] = `
<Fragment>
  <div
    className="spinner"
  >
    <div
      className="rect1"
    />
    <div
      className="rect2"
    />
    <div
      className="rect3"
    />
    <div
      className="rect4"
    />
    <div
      className="rect5"
    />
  </div>
  <div
    className="mask"
  >
    Children
  </div>
</Fragment>
`;
